### PR TITLE
Prevent a memory leak

### PR DESCRIPTION
**Pull request by pipiche38**
In relation with #5881 but not addressing the all issue

PyUnicode_FromString returns a new reference, then PyDict_SetItemString stores that reference and increfs, but the string isn’t decrefed.

Most-likely this must be done everywhere PyDict_SetItemString()

PyDict_SetItemString(self->Options, "Custom", PyUnicode_FromString(sd[14].c_str()));

PyUnicode_FromString returns a new reference, then PyDict_SetItemString stores that reference and increfs, but the string isn’t decrefed.

Suppose that PyUnicode_FromString returned a new string. Its refcount would be1.

PyDict_SetItemString would store that string and increment the refcount to 2.

### CODE REVIEW

**Comment by xbeaudouin** [comment]
This is interesting code. Will add that to my port... because I still have some leak... even if I don't use anymore zigbee plugin...

**Comment by pipiche38** [comment]
Xavier , that is wip... hold on

**Comment by xbeaudouin** [comment]
Keep me up to date... because really on a 8G machine... we should not have so much memory used....
Capture d’écran 2023-11-21 à 17 35 20

**Comment by dnpwwo** [comment]
This does look like a bug but can you use the smart pointer type rather than the raw Python objects for consistency:

Rather than:
```
	  PyObject* str = PyUnicode_FromString(sd[14].c_str());
	  PyDict_SetItemString(self->Options, "Custom", str);
	  Py_DECREF(str);
```
use:
```
	PyNewRef str = PyUnicode_FromString(sd[14].c_str());
	PyDict_SetItemString(self->Options, "Custom", str);
```

**Comment by pipiche38** [comment]
@dnpwwo , I did the change , not sure what it means. feel free to comment .

**Comment by dnpwwo** [comment]
You need to remove the 5 explicit `Py_DECREF(str);` statements otherwise unpredictable things will happen.

By using the smart pointer the reference counting is taken care of for you.  When the variable goes out of scope C++ guarantees that the count will be correctly decremented.

Python functions return to different pointer types:

1. Borrowed References.  These should not have their reference counts changed because you are borrowing something else's reference.
2. New References. These should have their reference counts decremented when you have finished with them because you have your own reference (these are the most common).

I created smart pointer types to match these are respect the rules

**Comment by mealgi** [comment]
Hello,

As i mentionned it is not always the same python plugin thread which causes the crash.

That is why I am not enclined to suspect  a dedicated python plugin among all the ones implemented.

BR

**Comment by pipiche38** [comment]
@mealgi I guess you commented the wrong post, as your comment as nothing to do with this PR.

**Comment by mealgi** [comment]
Hello
Sorry, my bad.
Yes I commented the wrong post.
I am also interested in the memroy leak issue
BR

**Comment by pipiche38** [comment]
@dnpwwo @gizmocuz I do consider that the PR is ready .

On one side, that fix few leaks related to Python plugin, but also to Python script.

However, I still see leakage via https://github.com/domoticz/domoticz/issues/5881 and I wonder if that could be linked to the framework itself . By doing the 100 Touch or Refresh per second, does the Garbage Collector has the possibility to take some action. As we are on the main stream, is the GC able to do action ? 

In conclusion, I recommend this PR for merge if you are happy with, and then we continue on the Issue #5881 and for that one, I would really need your help to be able to add GC calls , like for instance forcing the the GC.collect() to see if any impact or not.

**Comment by dnpwwo** [comment]
@pipiche38,

That doesn't surprise me.  The issues you found were real but they were to do with the Options field of the Unit.  The Units you created in your test script did not have options (Selector Switches and a few other types do) so I did not think they would resolve this problem.

PR looks okay to me

**Comment by xbeaudouin** [comment]
Will test that on my dev server. And then in my production platform... (current domoticz uses 4G of RES memory...)

**Comment by dnpwwo** [comment]
@pipiche38 

I've been running your plugin on Windows in Visual Studio and don't see any memory loss.  How are you provoking this?

**Comment by gizmocuz** [comment]
@dnpwwo , I also mentioned this above. I do not see a memory leak in windows/visual studio (before this PR)
It does increase a little bit, but keeps stable after a few seconds. (using python 3.12)

**Comment by xbeaudouin** [comment]
Patch is currently on production on my main domoticz : (FreeBSD 13.2 / python 3.9.18) with 16 python plugins running.

**Comment by pipiche38** [comment]
@gizmocuz @dnpwwo I answer your question in the issue itself https://github.com/domoticz/domoticz/issues/5881#issuecomment-1825693645

**Comment by xbeaudouin** [comment]
Well it seems FreeBSD gave some sanction:
```
Nov 24 17:31:14 domoticz domoticz[72164]: Domoticz(pid:72164, tid:0('-')) received fatal signal 11 (Segmentation fault)
Nov 24 17:31:14 domoticz domoticz[72164]: #0  0x7f3f45 <_ZL19dumpstack_backtracePvS_+0x35> at /usr/local/domoticz/domoticz
Nov 24 17:31:14 domoticz domoticz[72164]: #1  0x7f42c7 <_ZL9dumpstackPvS_+0x37> at /usr/local/domoticz/domoticz
Nov 24 17:31:14 domoticz domoticz[72164]: #2  0x7f3e62 <_Z14signal_handleriP9__siginfoPv+0x202> at /usr/local/domoticz/domoticz
Nov 24 17:31:14 domoticz domoticz[72164]: #3  0x826729b6e <pthread_sigmask+0x54e> at /lib/libthr.so.3
Nov 24 17:31:14 domoticz domoticz[72164]: #4  0x82672911f <pthread_setschedparam+0x83f> at /lib/libthr.so.3
Nov 24 17:31:14 domoticz domoticz[72164]: #5  0x7ffffffff2d3 <_ZNKSt3__114error_category10equivalentERKNS_10error_codeEi+0x7ffffef69333> at ???
Nov 24 17:31:14 domoticz domoticz[72164]: #6  0x8308a8dca <__sys_thr_kill+0xa> at /lib/libc.so.7
Nov 24 17:31:14 domoticz domoticz[72164]: #7  0x8308216f4 <__raise+0x34> at /lib/libc.so.7
Nov 24 17:31:14 domoticz domoticz[72164]: #8  0x826729b60 <pthread_sigmask+0x540> at /lib/libthr.so.3
Nov 24 17:31:14 domoticz domoticz[72164]: #9  0x82672911f <pthread_setschedparam+0x83f> at /lib/libthr.so.3
Nov 24 17:31:14 domoticz domoticz[72164]: #10 0x7ffffffff2d3 <_ZNKSt3__114error_category10equivalentERKNS_10error_codeEi+0x7ffffef69333> at ???
Nov 24 17:31:14 domoticz domoticz[72164]: #11 0x83b9ebcea <PyType_IsSubtype+0x769a> at /usr/local/lib/libpython3.9.so
Nov 24 17:31:14 domoticz domoticz[72164]: #12 0x83ba6b110 <_PyEval_EvalFrameDefault+0xb760> at /usr/local/lib/libpython3.9.so
Nov 24 17:31:14 domoticz domoticz[72164]: #13 0x83ba66f7e <_PyEval_EvalFrameDefault+0x75ce> at /usr/local/lib/libpython3.9.so
Nov 24 17:31:14 domoticz domoticz[72164]: #14 0x83b9787ed <_PyFunction_Vectorcall+0x21d> at /usr/local/lib/libpython3.9.so
Nov 24 17:31:14 domoticz domoticz[72164]: #15 0x83ba6ad27 <_PyEval_EvalFrameDefault+0xb377> at /usr/local/lib/libpython3.9.so
Nov 24 17:31:14 domoticz domoticz[72164]: #16 0x83ba66ecf <_PyEval_EvalFrameDefault+0x751f> at /usr/local/lib/libpython3.9.so
Nov 24 17:31:14 domoticz domoticz[72164]: #17 0x83b9787ed <_PyFunction_Vectorcall+0x21d> at /usr/local/lib/libpython3.9.so
Nov 24 17:31:14 domoticz domoticz[72164]: #18 0x83ba6ad27 <_PyEval_EvalFrameDefault+0xb377> at /usr/local/lib/libpython3.9.so
Nov 24 17:31:14 domoticz domoticz[72164]: #19 0x83ba66ecf <_PyEval_EvalFrameDefault+0x751f> at /usr/local/lib/libpython3.9.so
Nov 24 17:31:14 domoticz domoticz[72164]: #20 0x83b9787ed <_PyFunction_Vectorcall+0x21d> at /usr/local/lib/libpython3.9.so
Nov 24 17:31:14 domoticz domoticz[72164]: #21 0x83b9789fb <PyObject_CallObject+0xab> at /usr/local/lib/libpython3.9.so
Nov 24 17:31:14 domoticz domoticz[72164]: #22 0xf27983 <_ZN7Plugins7CPlugin8CallbackERNS_13PyBorrowedRefERKNSt3__112basic_stringIcNS3_11char_traitsIcEENS3_9allocatorIcEEEEP7_object+0x353> at /usr/local/domoticz/domoticz
Nov 24 17:31:14 domoticz domoticz[72164]: #23 0x6fcd88 <_ZN7Plugins13CCallbackBase8CallbackEPNS_7CPluginEP7_object+0xa8> at /usr/local/domoticz/domoticz
Nov 24 17:31:14 domoticz domoticz[72164]: #24 0xf310d5 <_ZN7Plugins19onHeartbeatCallback13ProcessLockedEPNS_7CPluginE+0x25> at /usr/local/domoticz/domoticz
Nov 24 17:31:14 domoticz domoticz[72164]: #25 0x6fcc95 <_ZN7Plugins18CPluginMessageBase7ProcessEPNS_7CPluginE+0x55> at /usr/local/domoticz/domoticz
Nov 24 17:31:14 domoticz domoticz[72164]: #26 0xf1c28d <_ZN7Plugins7CPlugin7Do_WorkEv+0x3fd> at /usr/local/domoticz/domoticz
Nov 24 17:31:14 domoticz domoticz[72164]: #27 0xf361c9 <_ZNSt3__18__invokeIMN7Plugins7CPluginEFvvEPS2_JEvEEDTcldsdescT0_fp0_fp_spscT1_fp1_EEOT_OS6_DpOS7_+0x69> at /usr/local/domoticz/domoticz
Nov 24 17:31:14 domoticz domoticz[72164]: #28 0xf3614e <_ZNSt3__116__thread_executeINS_10unique_ptrINS_15__thread_structENS_14default_deleteIS2_EEEEMN7Plugins7CPluginEFvvEJPS7_EJLm2EEEEvRNS_5tupleIJT_T0_DpT1_EEENS_15__tu
ple_indicesIJXspT2_EEEE+0x3e> at /usr/local/domoticz/domoticz
Nov 24 17:31:14 domoticz domoticz[72164]: #29 0xf35e4d <_ZNSt3__114__thread_proxyINS_5tupleIJNS_10unique_ptrINS_15__thread_structENS_14default_deleteIS3_EEEEMN7Plugins7CPluginEFvvEPS8_EEEEEPvSD_+0x6d> at /usr/local/domoticz/domoticz
```

**Comment by pipiche38** [comment]
@xbeaudouin I don't if that is related. I guess @gizmocuz or @dnpwwo  can comment.
On my end this is beyond my expertise level. So either we close this PR or someone else take it over to fix.

**Comment by gizmocuz** [comment]
@pipiche38 , I do think this is related because it mentions the shared pointers.
I also think that this PR does not solve the problem. You already mentioned this as well in the Issue when you just returned in the function and not doing anything

**Comment by dnpwwo** [comment]
From experience, changing the reference counting (particularly when you decrement counters more) often has unexpected side effects that can show up later.  If the reference count drops to zero on an object the GC will free the memory but if there is another reference to the object anywhere else the code will crash later when the reference is used.

If @xbeaudouin had a stable system before and only applied this PR and now has a crash then it is definitely related.

**Comment by xbeaudouin** [comment]
Well... I don't have any memory leak... because domoticz sig11 almost every day :(
With almost same debug info as I put on the PR that was closed.
Also I notice that _Py_Dealloc() is removed on Python 3.9 and more... With the include we seems to "make it" running again, maybe we will have to see deeper in python embedded API...
Unfortunatly I dunno enougth python for that... :(
